File: frontend/src/types/alerts.ts
export type AlertMetric = "temperature" | "humidity";
export type AlertOperator = "greater_than" | "less_than";
export type AlertSeverity = "info" | "warning" | "critical";
export type AlertRuntimeState = "normal" | "pending" | "active";
export type AlertEventStatus = "active" | "resolved";

export interface AlertRuleInput {
    name: string;
    device_id: string;
    metric: AlertMetric;
    operator: AlertOperator;
    threshold: number;
    duration_seconds: number;
    hysteresis: number;
    severity: AlertSeverity;
    enabled: boolean;
}

export interface AlertRule extends AlertRuleInput {
    id: number;
    created_at: string;
    updated_at: string;
    archived_at: string | null;
    runtime_state: AlertRuntimeState;
    pending_since: string | null;
    last_evaluated_at: string | null;
    last_value: number | null;
}

export interface AlertEvent {
    id: number;
    rule_id: number;
    device_id: string;
    rule_name: string;
    metric: AlertMetric;
    operator: AlertOperator;
    threshold: number;
    hysteresis: number;
    duration_seconds: number;
    severity: AlertSeverity;
    status: AlertEventStatus;
    started_at: string;
    activated_at: string;
    resolved_at: string | null;
    resolution_reason: string | null;
    activation_value: number;
    last_value: number;
    extreme_value: number;
}

File: backend/domain/alerts.py
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Literal


AlertMetric = Literal["temperature", "humidity"]
AlertOperator = Literal["greater_than", "less_than"]
AlertSeverity = Literal["info", "warning", "critical"]
RuleRuntimeState = Literal["normal", "pending", "active"]
AlertEventStatus = Literal["active", "resolved"]
ResolutionReason = Literal[
    "condition_recovered",
    "rule_disabled",
    "rule_updated",
    "rule_archived",
]


@dataclass(frozen=True)
class RuleCondition:
    metric: AlertMetric
    operator: AlertOperator
    threshold: float
    hysteresis: float


def is_violation(condition: RuleCondition, value: float) -> bool:
    if condition.operator == "greater_than":
        return value > condition.threshold
    return value < condition.threshold


def is_recovery(condition: RuleCondition, value: float) -> bool:
    if condition.operator == "greater_than":
        return value <= condition.threshold - condition.hysteresis
    return value >= condition.threshold + condition.hysteresis


def update_extreme(operator: AlertOperator, current: float, value: float) -> float:
    return max(current, value) if operator == "greater_than" else min(current, value)


def telemetry_metric_value(sample: dict, metric: AlertMetric) -> float:
    if metric == "temperature":
        return float(sample["temperature"])
    if metric == "humidity":
        return float(sample["humidity"])
    raise ValueError(f"unsupported alert metric: {metric}")


def parse_utc_timestamp(value: str) -> datetime:
    parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if parsed.tzinfo is None or parsed.utcoffset() is None:
        raise ValueError("timestamp must be timezone-aware")
    return parsed.astimezone(timezone.utc)


def utc_isoformat(value: datetime) -> str:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("timestamp must be timezone-aware")
    return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")
